roomserver/internal/perform/perform_upgrade.go

```diff
@@ -0,0 +1,691 @@
+			Msg: "Power level event content was invalid",
+		}
+	}
+
+	// We might need to temporarily give ourselves a higher power level
+	// than we had in the old room in order to be able to send all of
+	// the relevant state events. This function will return whether we
+	// had to override the power level events or not — if we did, we
+	// need to send the original power levels again later on.
+	tempPowerLevelsEvent, powerLevelsOverridden := createTemporaryPowerLevels(powerLevelContent, userID)
+
+	joinRulesContent, err := oldJoinRulesEvent.JoinRule()
+	if err != nil {
+		return nil, &api.PerformError{
+			Msg: "Join rules event content was invalid",
+		}
+	}
+	newJoinRulesEvent := fledglingEvent{
+		Type: gomatrixserverlib.MRoomJoinRules,
+		Content: map[string]interface{}{
+			"join_rule": joinRulesContent,
+		},
+	}
+
+	eventsToMake := make([]fledglingEvent, 0, len(state))
+	eventsToMake = append(eventsToMake, newCreateEvent, membershipEvent, tempPowerLevelsEvent, newJoinRulesEvent)
+
+	if _, ok := state[gomatrixserverlib.StateKeyTuple{EventType: gomatrixserverlib.MRoomGuestAccess, StateKey: ""}]; !ok {
+		// Appease sytest, as it expects a guest access event for some reason.
+		eventsToMake = append(eventsToMake, fledglingEvent{
+			Type: gomatrixserverlib.MRoomGuestAccess,
+			Content: map[string]string{
+				"guest_access": "forbidden",
+			},
+		})
+	}
+
+	// Duplicate all of the old state events into the new room.
+	for tuple, event := range state {
+		if _, ok := override[tuple]; ok {
+			// Don't duplicate events we have overridden already. They
+			// are already in `eventsToMake`.
+			continue
+		}
+		if membership, merr := event.Membership(); merr == nil && membership != gomatrixserverlib.Ban {
+			// Don't duplicate membership events that aren't bans. Our own
+			// membership event has already been created above, and event auth
+			// won't let us create join membership events for other users.
+			continue
+		}
+		newEvent := fledglingEvent{
+			Type:     tuple.EventType,
+			StateKey: tuple.StateKey,
+		}
+		if err = json.Unmarshal(event.Content(), &newEvent.Content); err != nil {
+			logrus.WithError(err).Error("Failed to unmarshal old event")
+			continue
+		}
+		eventsToMake = append(eventsToMake, newEvent)
+	}
+
+	// If we sent a temporary power level event into the room before,
+	// override that now by restoring the original power levels.
+	if powerLevelsOverridden {
+		eventsToMake = append(eventsToMake, fledglingEvent{
+			Type:    gomatrixserverlib.MRoomPowerLevels,
+			Content: powerLevelContent,
+		})
+	}
+	return eventsToMake, nil
+}
+
+func (r *Upgrader) sendInitialEvents(ctx context.Context, evTime time.Time, userID, newRoomID, newVersion string, eventsToMake []fledglingEvent) *api.PerformError {
+	var err error
+	var builtEvents []*gomatrixserverlib.HeaderedEvent
+	authEvents := gomatrixserverlib.NewAuthEvents(nil)
+	for i, e := range eventsToMake {
+		depth := i + 1 // depth starts at 1
```

Comment:
Surely this sort of stuff should be shoved into GMSL or eventutil? We do the same thing in /createRoom.

Reply:
Possibly, but then we also have to refactor that part of `/createRoom`, which feels like it's sprawling out of the scope of this PR.